hv/crates/hv-cell/src/lib.rs

```diff
@@ -178,6 +179,75 @@ impl<T: ?Sized> ArcCell<T> {
     pub fn from_inner(inner: Arc<AtomicRefCell<T>>) -> Self {
         Self { inner }
     }
+
+    /// Construct a `WeakCell<T>`
+    pub fn downgrade(this: &Self) -> WeakCell<T> {
+        WeakCell {
+            inner: Arc::downgrade(&this.inner),
+        }
+    }
+}
+
+/// An analogue to `Arc<T>`'s `Weak<T>` type, normally constructed via `ArcCell::downgrade(this: &T)` or `WeakCell::new()`
+pub struct WeakCell<T: ?Sized> {
+    inner: Weak<AtomicRefCell<T>>,
+}
+
+impl<T> Default for WeakCell<T> {
+    fn default() -> Self {
+        WeakCell::new()
+    }
+}
+
+impl<T> WeakCell<T> {
+    #[inline]
+    /// Construct a new `WeakCell<T>`
+    pub fn new() -> Self {
+        WeakCell { inner: Weak::new() }
+    }
+
+    /// Retrieve the inner field of the `WeakCell<T>`
+    pub fn as_inner(&self) -> &Weak<AtomicRefCell<T>> {
+        &self.inner
+    }
+
+    /// Unsafely try to borrow the underlying `T` (note that this could panic either due to a `borrow`
+    /// error, or due to an `upgrade` error)
+    pub fn borrow(&self) -> ArcRef<T> {
+        self.upgrade().unwrap().borrow()
+    }
+
+    /// Unsafely try to mutably borrow the underlying `T` (note that this could panic either due to a `borrow_mut`
+    /// error, or due to an `upgrade` error)
+    pub fn borrow_mut(&mut self) -> ArcRefMut<T> {
+        self.upgrade().unwrap().borrow_mut()
+    }
+
+    /// Safely attempt to borrow the underlying `T`. It first attempts to upgrade the `WeakCell<T>` before
```

Comment:
There are no methods on `Weak` which could possibly be implemented unsafely in the first place, so safety is not relevant here. That it will not panic is notable, but distinct.